lib/grandpa/grandpa.go

```diff
@@ -38,178 +50,192 @@ func NewService(blockState BlockState, voters []*Voter) (*Service, error) {
-	vote := NewVoteFromHeader(header)
-
-	msg, err := scale.Encode(&FullVote{
-		stage: s.subround,
-		vote:  vote,
-		round: s.state.round,
-		setID: s.state.setID,
-	})
+// getPreVotedBlock returns the current pre-voted block B.
+// the pre-voted block is the block with the highest block number in the set of all the blocks with
+// total votes >= 2/3 the total number of voters, where the total votes is determined by getTotalVotesForBlock.
+func (s *Service) getPreVotedBlock() (Vote, error) {
+	blocks, err := s.getPossiblePreVotedBlocks()
 	if err != nil {
-		return nil, err
+		return Vote{}, err
 	}
 
-	sig, err := kp.Sign(msg)
-	if err != nil {
-		return nil, err
+	if len(blocks) == 0 {
+		return Vote{}, ErrNoPreVotedBlock
 	}
 
-	sm := &SignedMessage{
-		hash:        vote.hash,
-		number:      vote.number,
-		signature:   ed25519.NewSignatureBytes(sig),
-		authorityID: kp.Public().(*ed25519.PublicKey).AsBytes(),
+	// if there is one block, return it
+	if len(blocks) == 1 {
+		for h, n := range blocks {
+			return Vote{
+				hash:   h,
+				number: n,
+			}, nil
+		}
 	}
 
-	return &VoteMessage{
-		setID:   s.state.setID,
-		round:   s.state.round,
-		stage:   s.subround,
-		message: sm,
-	}, nil
-}
-
-// ValidateMessage validates a VoteMessage and adds it to the current votes
-// it returns the resulting vote if validated, error otherwise
-func (s *Service) ValidateMessage(m *VoteMessage) (*Vote, error) {
-	// check for message signature
-	pk, err := ed25519.NewPublicKey(m.message.authorityID[:])
-	if err != nil {
-		return nil, err
+	// if there are multiple, find the one with the highest number and return it
+	highest := Vote{
+		number: uint64(0),
 	}
-
-	err = validateMessageSignature(pk, m)
-	if err != nil {
-		return nil, err
+	for h, n := range blocks {
+		if n > highest.number {
+			highest = Vote{
+				hash:   h,
+				number: n,
+			}
+		}
 	}
 
-	// check that setIDs match
-	if m.setID != s.state.setID {
-		return nil, ErrSetIDMismatch
+	return highest, nil
+}
+
+// getPossiblePreVotedBlocks returns blocks with total votes >=2/3 the total number of voters in the map of block hash to block number.
+// if there are no blocks that have >=2/3 direct votes, this function will find predecessors of those blocks that do have >=2/3 votes.
+// note that by voting for a block, all of its predecessor blocks are automatically voted for.
+// thus, if there are no blocks with >=2/3 total votes, but the sum of votes for blocks A and B is >=2/3, then this function returns
+// the first common predecessor of A and B.
+func (s *Service) getPossiblePreVotedBlocks() (map[common.Hash]uint64, error) {
+	// get blocks that were directly voted for
+	votes := s.getDirectVotes()
+	blocks := make(map[common.Hash]uint64)
+
+	// check if any of them have >=2/3 votes
+	for v := range votes {
+		total, err := s.getTotalVotesForBlock(v.hash)
+		if err != nil {
+			return nil, err
+		}
+
+		if total >= uint64(2*len(s.state.voters)/3) {
```

Comment:
the condition could be more clear if the `2*len(s.state.voters)/3` would be extracted to a var that would represent the operation in a more meaningful way.

Reply:
changed to `s.state.threshold()`